### ISSUE TITLE
MissingDependencyException encountered when trying to remove a firebase dependency

### ISSUE THREAD

**Issue by douglasselph** [firebase/firebase-android-sdk#2605]
* Android Studio version: 4.1.3
  * Firebase Component: Not sure, see below
  * Component version: Not sure, see below
 
#### Steps to reproduce:

First we have had our app running with these dependencies for quite a while:

```
    implementation 'com.google.firebase:firebase-analytics-ktx:18.0.3'
    implementation 'com.google.firebase:firebase-auth-ktx:20.0.4'
    implementation 'com.google.firebase:firebase-config-ktx:20.0.4'
    implementation 'com.google.firebase:firebase-crashlytics-ktx:17.4.1'
    implementation 'com.google.firebase:firebase-database-ktx:19.7.0'
    implementation 'com.google.firebase:firebase-dynamic-links-ktx:19.1.1'
    implementation 'com.google.firebase:firebase-inappmessaging-display-ktx:19.1.5'
    implementation 'com.google.firebase:firebase-messaging-ktx:21.0.1'
    implementation 'com.google.firebase:firebase-storage-ktx:19.2.2'
    implementation 'com.google.firebase:firebase-firestore-ktx:22.1.2'
    implementation 'com.google.firebase:firebase-perf:19.1.1'
    implementation 'com.google.android.gms:play-services-mlkit-barcode-scanning:16.1.4'
    implementation 'com.google.mlkit:smart-reply:16.1.1'
```

   Then because I wanted to make the smart-reply component a module, I decided to move the related smart reply code to a module. Then I was able to remove this above dependency from our main app build.gradle file:

`    implementation 'com.google.mlkit:smart-reply:16.1.1'
`

Without this, the code compiled fine. However, when I tried to run the app I got this crash during start up:

```
D FirebaseAuth: Notifying id token listeners about user ( GLhOauaumnQzSQVVRZs056zMsd23 ).
I FirebaseInitProvider: FirebaseApp initialization successful
I FirebaseAuth: [FirebaseAuth:] Preparing to create service connection to fallback implementation
D AndroidRuntime: Shutting down VM
E AndroidRuntime: FATAL EXCEPTION: main
E AndroidRuntime: Process: com.callassistant.android, PID: 21283
E AndroidRuntime: java.lang.RuntimeException: Unable to get provider com.google.mlkit.common.internal.MlKitInitProvider: com.google.firebase.components.MissingDependencyException: Unsatisfied dependency for component Component<[class com.google.android.gms.internal.mlkit_naturallanguage.zzfz]>{0, type=0, deps=[Dependency{anInterface=class com.google.mlkit.common.sdkinternal.SharedPrefManager, type=required, injection=direct}, Dependency{anInterface=class android.content.Context, type=required, injection=direct}, Dependency{anInterface=interface com.google.android.gms.internal.mlkit_naturallanguage.zzfz$zza, type=required, injection=direct}]}: class com.google.mlkit.common.sdkinternal.SharedPrefManager
E AndroidRuntime: 	at android.app.ActivityThread.installProvider(ActivityThread.java:7247)
E AndroidRuntime: 	at android.app.ActivityThread.installContentProviders(ActivityThread.java:6783)
E AndroidRuntime: 	at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6700)
E AndroidRuntime: 	at android.app.ActivityThread.access$1300(ActivityThread.java:237)
E AndroidRuntime: 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1913)
E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:106)
E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:223)
E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:7660)
E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
E AndroidRuntime: Caused by: com.google.firebase.components.MissingDependencyException: Unsatisfied dependency for component Component<[class com.google.android.gms.internal.mlkit_naturallanguage.zzfz]>{0, type=0, deps=[Dependency{anInterface=class com.google.mlkit.common.sdkinternal.SharedPrefManager, type=required, injection=direct}, Dependency{anInterface=class android.content.Context, type=required, injection=direct}, Dependency{anInterface=interface com.google.android.gms.internal.mlkit_naturallanguage.zzfz$zza, type=required, injection=direct}]}: class com.google.mlkit.common.sdkinternal.SharedPrefManager
E AndroidRuntime: 	at com.google.firebase.components.ComponentRuntime.processDependencies(ComponentRuntime.java:324)
E AndroidRuntime: 	at com.google.firebase.components.ComponentRuntime.discoverComponents(ComponentRuntime.java:139)
E AndroidRuntime: 	at com.google.firebase.components.ComponentRuntime.<init>(ComponentRuntime.java:91)
E AndroidRuntime: 	at com.google.firebase.components.ComponentRuntime.<init>(ComponentRuntime.java:65)
E AndroidRuntime: 	at com.google.mlkit.common.sdkinternal.MlKitContext.zza(com.google.mlkit:common@@17.1.1:6)
E AndroidRuntime: 	at com.google.mlkit.common.internal.MlKitInitProvider.onCreate(com.google.mlkit:common@@17.1.1:3)
E AndroidRuntime: 	at android.content.ContentProvider.attachInfo(ContentProvider.java:2388)
E AndroidRuntime: 	at android.content.ContentProvider.attachInfo(ContentProvider.java:2358)
E AndroidRuntime: 	at com.google.mlkit.common.internal.MlKitInitProvider.attachInfo(com.google.mlkit:common@@17.1.1:3)
E AndroidRuntime: 	at android.app.ActivityThread.installProvider(ActivityThread.java:7242)
E AndroidRuntime: 	... 10 more
V FA      : Connecting to remote service
```

#### Relevant Code:

If the above is not clear enough, or if you want me to try and create reproducible scenario then I could try to reproduce it by first running the tutorial app with https://github.com/googlesamples/mlkit/tree/master/android/smartreply; and then remove all the firebase related code and see if it still runs. But to be honest I don't want to spend the time doing so unless you can see absolutely nothing wrong with the code referenced in stack trace above.

It seems to me based on the stack trace, that firebase is forcing a dependency on mlkit because it "remembered" somehow that the app used mlkit previously. Is that so?

I can make the crash go away be re-inserting the com.google.mlkit:smart-reply:16.1.1 dependency. But I don't want to do that since I am no longer referencing it anywhere in our code.

**Comment by cs-googler** [comment]
Hi, play store feature module is not supported by the Firebase ML Kit. 

We made some changes to Firebase ML Kit to better distinguish the on-device APIs from cloud based APIs. "[ML Kit](https://developers.google.com/ml-kit/guides)"(without firebase branding) contains all the on-device APIs. [Here's]( https://developers.google.com/ml-kit/migration) the migration guide from firebase mlkit to mlkit. 

We will add the support for play store feature module soon in the standalone MLKit. All further improvements and new APIs will be released only with the new ML Kit. Please stay tuned.

**Comment by douglasselph** [comment]
Okay, I will study what you sent me when I get the chance. 

Just in case it is helpful, I did find another way to get it working:
I moved all my code that referenced:
`   implementation 'com.google.android.gms:play-services-mlkit-barcode-scanning:16.1.4'
    implementation 'com.google.mlkit:smart-reply:16.1.1'`
into my module and it compiled and ran fine. That is, if I removed all both mlkit libraries out it works. But moving just one did not. Apparently, if you have both of these together (barcode & smartreply) and then subsequently remove smartreply, it creates this issue. Thanks.
